api/scpca_portal/models/sample.py

```diff
@@ -257,9 +257,34 @@ def create_computed_files(
     ):
         """Prepares ready for saving project computed files based on generated file mappings."""
 
+        # Prepare a threading.Lock for each sample, with the chief purpose being to protect
+        # multiplexed samples that shares a zip file.
+        locks = {}
+
         def on_get_sample_file(future):
             if computed_file := future.result():
-                computed_file.process_computed_file(clean_up_output_data, update_s3)
+                # Once created, save immediately, so that it can be queried below
+                computed_file.save()
+
+                config = {
+                    "modality": computed_file.modality,
+                    "format": computed_file.format,
+                }
+                # Protect from duplicate computed file uploads with multiplexed samples
+                with locks.get(sample.get_config_identifier(config)):
+                    nonlocal update_s3
+                    if update_s3:
+                        computed_file.upload_s3
+                        update_s3 = False
+
+                all_samples_created = (
+                    len(computed_file.sample.multiplexed_ids)
+                    == ComputedFile.objects.filter(s3_key=computed_file.s3_key).count()
+                )
```

Comment:
As a general aside, this is the type of intermediate querying that could justify a Public=False or Loading=True attribute on the ComputedFile model. Not saying that that's the route that we need to take, but I guess something like this is what I had in mind.

Reply:
```python
        def on_get_sample_file(future):
            if computed_file := future.result():
                # Once created, save immediately, so that it can be queried below

                # Only upload and clean up the last multiplexed sample
                if computed_file.sample.is_last_sample:
                    if update_s3:
                        computed_file.upload_s3_file()
                    if clean_up_output_data:
                        computed_file.clean_up_local_computed_file()

                computed_file.save()
            # Close DB connection for each thread.
            connection.close()

```

Reply:
No need for a lock, we can just check if it is the last sample and then upload it / clean up. Otherwise, just save and bypass those steps.

Reply:
- Add `Sample::is_last_sample` `@property` method
- Also move `locks` back down below